File: src/theme.zig
const std = @import("std");
const util = @import("util.zig");

pub const ThemeError = error{
    ThemeInvalid,
} || std.mem.Allocator.Error;

pub const Color = struct {
    hex: []const u8,
};

pub const ThemeTokens = struct {
    name: []const u8,
    bg_default: Color,
    bg_panel: Color,
    bg_selected: Color,
    fg_default: Color,
    fg_muted: Color,
    fg_accent: Color,
    diff_add_bg: Color,
    diff_add_fg: Color,
    diff_del_bg: Color,
    diff_del_fg: Color,
    diff_context_fg: Color,
    border: Color,
    warning: Color,
    error_color: Color,
    comment_badge: Color,
    reviewed_badge: Color,
    unreviewed_badge: Color,
    syntax_keyword: Color,
    syntax_string: Color,
    syntax_comment: Color,
    syntax_type: Color,
    syntax_function: Color,
    syntax_number: Color,
    syntax_operator: Color,
    syntax_plain: Color,
};

pub fn catppuccinMocha() ThemeTokens {
    return .{
        .name = "catppuccin-mocha",
        .bg_default = .{ .hex = "#1e1e2e" },
        .bg_panel = .{ .hex = "#181825" },
        .bg_selected = .{ .hex = "#313244" },
        .fg_default = .{ .hex = "#cdd6f4" },
        .fg_muted = .{ .hex = "#9399b2" },
        .fg_accent = .{ .hex = "#89b4fa" },
        .diff_add_bg = .{ .hex = "#1e3a2f" },
        .diff_add_fg = .{ .hex = "#a6e3a1" },
        .diff_del_bg = .{ .hex = "#3b2228" },
        .diff_del_fg = .{ .hex = "#f38ba8" },
        .diff_context_fg = .{ .hex = "#bac2de" },
        .border = .{ .hex = "#45475a" },
        .warning = .{ .hex = "#f9e2af" },
        .error_color = .{ .hex = "#f38ba8" },
        .comment_badge = .{ .hex = "#f5c2e7" },
        .reviewed_badge = .{ .hex = "#a6e3a1" },
        .unreviewed_badge = .{ .hex = "#fab387" },
        .syntax_keyword = .{ .hex = "#cba6f7" },
        .syntax_string = .{ .hex = "#a6e3a1" },
        .syntax_comment = .{ .hex = "#6c7086" },
        .syntax_type = .{ .hex = "#f9e2af" },
        .syntax_function = .{ .hex = "#89b4fa" },
        .syntax_number = .{ .hex = "#fab387" },
        .syntax_operator = .{ .hex = "#94e2d5" },
        .syntax_plain = .{ .hex = "#cdd6f4" },
    };
}

pub const Ansi = struct {
    enabled: bool,
    true_color: bool,

    pub fn init(force: bool) Ansi {
        if (!force and util.envExists("NO_COLOR")) return .{ .enabled = false, .true_color = false };
        const term = (util.envOwned(std.heap.page_allocator, "COLORTERM") catch null) orelse return .{ .enabled = true, .true_color = false };
        defer std.heap.page_allocator.free(term);
        return .{ .enabled = true, .true_color = util.contains(term, "truecolor") or util.contains(term, "24bit") };
    }

    pub fn fg(self: Ansi, allocator: std.mem.Allocator, color: Color) ![]u8 {
        if (!self.enabled) return util.dupe(allocator, "");
        if (!self.true_color) return util.dupe(allocator, "");
        const rgb = parseHex(color.hex) orelse return util.dupe(allocator, "");
        return std.fmt.allocPrint(allocator, "\x1b[38;2;{d};{d};{d}m", .{ rgb.r, rgb.g, rgb.b });
    }

    pub fn bg(self: Ansi, allocator: std.mem.Allocator, color: Color) ![]u8 {
        if (!self.enabled) return util.dupe(allocator, "");
        if (!self.true_color) return util.dupe(allocator, "");
        const rgb = parseHex(color.hex) orelse return util.dupe(allocator, "");
        return std.fmt.allocPrint(allocator, "\x1b[48;2;{d};{d};{d}m", .{ rgb.r, rgb.g, rgb.b });
    }

    pub fn reset(self: Ansi) []const u8 {
        return if (self.enabled) "\x1b[0m" else "";
    }

    /// Write the foreground escape into a caller-owned buffer and return the slice (empty when
    /// colour is disabled). Avoids the per-call heap allocation that fg() incurs on the hot
    /// per-cell / per-span render path. The returned slice borrows `buf`.
    pub fn fgInto(self: Ansi, buf: *[24]u8, color: Color) []const u8 {
        return self.sgrColorInto(buf, "38", color);
    }

    pub fn bgInto(self: Ansi, buf: *[24]u8, color: Color) []const u8 {
        return self.sgrColorInto(buf, "48", color);
    }

    fn sgrColorInto(self: Ansi, buf: *[24]u8, comptime plane: []const u8, color: Color) []const u8 {
        if (!self.enabled or !self.true_color) return "";
        const rgb = parseHex(color.hex) orelse return "";
        return std.fmt.bufPrint(buf, "\x1b[" ++ plane ++ ";2;{d};{d};{d}m", .{ rgb.r, rgb.g, rgb.b }) catch "";
    }

    /// Append the foreground escape directly to `out`, allocation-free.
    pub fn appendFg(self: Ansi, allocator: std.mem.Allocator, out: *std.ArrayList(u8), color: Color) !void {
        var buf: [24]u8 = undefined;
        try out.appendSlice(allocator, self.fgInto(&buf, color));
    }
};

const Rgb = struct { r: u8, g: u8, b: u8 };

fn parseHex(hex: []const u8) ?Rgb {
    const raw = if (hex.len > 0 and hex[0] == '#') hex[1..] else hex;
    if (raw.len != 6) return null;
    return .{
        .r = std.fmt.parseInt(u8, raw[0..2], 16) catch return null,
        .g = std.fmt.parseInt(u8, raw[2..4], 16) catch return null,
        .b = std.fmt.parseInt(u8, raw[4..6], 16) catch return null,
    };
}

pub fn validateBaseThemeFile(allocator: std.mem.Allocator, io: std.Io, path: []const u8) ThemeError![]u8 {
    const content = std.Io.Dir.cwd().readFileAlloc(io, path, allocator, .limited(4 * 1024 * 1024)) catch return error.ThemeInvalid;
    defer allocator.free(content);
    var base16_count: usize = 0;
    var base24_count: usize = 0;
    var idx: u8 = 0;
    while (idx < 24) : (idx += 1) {
        var key_buf: [6]u8 = undefined;
        const key = std.fmt.bufPrint(&key_buf, "base{x:0>2}", .{idx}) catch unreachable;
        // Require each baseNN key to carry a real 6-hex colour on its line, not merely appear as
        // a substring, so arbitrary files that happen to contain "base00" are not accepted.
        if (!keyHasHexColor(content, key)) continue;
        if (idx < 16) base16_count += 1 else base24_count += 1;
    }
    if (base16_count < 16) return error.ThemeInvalid;
    return std.fmt.allocPrint(allocator, "valid Base{s} theme ({d} color slots)", .{ if (base24_count >= 8) "24" else "16", base16_count + base24_count });
}

fn keyHasHexColor(content: []const u8, key: []const u8) bool {
    var search_from: usize = 0;
    while (std.mem.indexOfPos(u8, content, search_from, key)) |key_idx| {
        const after = content[key_idx + key.len ..];
        const line_end = std.mem.indexOfScalar(u8, after, '\n') orelse after.len;
        if (lineContainsHexColor(after[0..line_end])) return true;
        search_from = key_idx + key.len;
    }
    return false;
}

fn lineContainsHexColor(line: []const u8) bool {
    var i: usize = 0;
    while (i + 6 <= line.len) : (i += 1) {
        const before_ok = i == 0 or !std.ascii.isHex(line[i - 1]);
        const after_ok = i + 6 == line.len or !std.ascii.isHex(line[i + 6]);
        if (before_ok and after_ok and parseHex(line[i .. i + 6]) != null) return true;
    }
    return false;
}

pub fn listBuiltins(allocator: std.mem.Allocator) ![]u8 {
    return std.fmt.allocPrint(allocator,
        \\catppuccin-mocha  built-in default
        \\base16/base24     supported via diffo themes validate <file>
        \\
    , .{});
}

test "base16 validator accepts keys" {
    const allocator = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    var content: std.ArrayList(u8) = .empty;
    defer content.deinit(allocator);
    var idx: u8 = 0;
    while (idx < 16) : (idx += 1) {
        var key_buf: [6]u8 = undefined;
        const key = std.fmt.bufPrint(&key_buf, "base{x:0>2}", .{idx}) catch unreachable;
        try content.appendSlice(allocator, key);
        try content.appendSlice(allocator, ": \"ffffff\"\n");
    }
    try tmp.dir.writeFile(std.testing.io, .{ .sub_path = "theme.yaml", .data = content.items });
    const tmp_path_z = try tmp.dir.realPathFileAlloc(std.testing.io, "theme.yaml", allocator);
    const tmp_path: []const u8 = tmp_path_z;
    defer allocator.free(tmp_path_z);
    const result = try validateBaseThemeFile(allocator, std.testing.io, tmp_path);
    defer allocator.free(result);
    try std.testing.expect(util.contains(result, "Base16"));
}

File: src/theme.rs
use crate::{Error, Result};
use std::{fs::File, io::Read, path::Path};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Color {
    pub hex: &'static str,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct ThemeTokens {
    pub name: &'static str,
    pub bg_default: Color,
    pub bg_panel: Color,
    pub bg_selected: Color,
    pub fg_default: Color,
    pub fg_muted: Color,
    pub fg_accent: Color,
    pub diff_add_bg: Color,
    pub diff_add_fg: Color,
    pub diff_del_bg: Color,
    pub diff_del_fg: Color,
    pub diff_context_fg: Color,
    pub border: Color,
    pub warning: Color,
    pub error_color: Color,
    pub comment_badge: Color,
    pub reviewed_badge: Color,
    pub unreviewed_badge: Color,
    pub syntax_keyword: Color,
    pub syntax_string: Color,
    pub syntax_comment: Color,
    pub syntax_type: Color,
    pub syntax_function: Color,
    pub syntax_number: Color,
    pub syntax_operator: Color,
    pub syntax_plain: Color,
}

pub fn catppuccin_mocha() -> ThemeTokens {
    ThemeTokens {
        name: "catppuccin-mocha",
        bg_default: Color { hex: "#1e1e2e" },
        bg_panel: Color { hex: "#181825" },
        bg_selected: Color { hex: "#313244" },
        fg_default: Color { hex: "#cdd6f4" },
        fg_muted: Color { hex: "#9399b2" },
        fg_accent: Color { hex: "#89b4fa" },
        diff_add_bg: Color { hex: "#1e3a2f" },
        diff_add_fg: Color { hex: "#a6e3a1" },
        diff_del_bg: Color { hex: "#3b2228" },
        diff_del_fg: Color { hex: "#f38ba8" },
        diff_context_fg: Color { hex: "#bac2de" },
        border: Color { hex: "#45475a" },
        warning: Color { hex: "#f9e2af" },
        error_color: Color { hex: "#f38ba8" },
        comment_badge: Color { hex: "#f5c2e7" },
        reviewed_badge: Color { hex: "#a6e3a1" },
        unreviewed_badge: Color { hex: "#fab387" },
        syntax_keyword: Color { hex: "#cba6f7" },
        syntax_string: Color { hex: "#a6e3a1" },
        syntax_comment: Color { hex: "#6c7086" },
        syntax_type: Color { hex: "#f9e2af" },
        syntax_function: Color { hex: "#89b4fa" },
        syntax_number: Color { hex: "#fab387" },
        syntax_operator: Color { hex: "#94e2d5" },
        syntax_plain: Color { hex: "#cdd6f4" },
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Ansi {
    pub enabled: bool,
    pub true_color: bool,
}

impl Ansi {
    pub fn init(force: bool) -> Self {
        let no_color = std::env::var_os("NO_COLOR").is_some();
        let colorterm = std::env::var_os("COLORTERM");
        Self::from_environment(
            force,
            no_color,
            colorterm
                .as_ref()
                .map(|value| value.to_string_lossy())
                .as_deref(),
        )
    }

    fn from_environment(force: bool, no_color: bool, colorterm: Option<&str>) -> Self {
        if !force && no_color {
            return Self {
                enabled: false,
                true_color: false,
            };
        }
        Self {
            enabled: true,
            true_color: colorterm
                .is_some_and(|value| value.contains("truecolor") || value.contains("24bit")),
        }
    }

    pub fn fg(self, color: Color) -> String {
        self.sgr_color(38, color)
    }

    pub fn bg(self, color: Color) -> String {
        self.sgr_color(48, color)
    }

    pub fn reset(self) -> &'static str {
        if self.enabled { "\x1b[0m" } else { "" }
    }

    fn sgr_color(self, plane: u8, color: Color) -> String {
        if !self.enabled || !self.true_color {
            return String::new();
        }
        match parse_hex(color.hex.as_bytes()) {
            Some([r, g, b]) => format!("\x1b[{plane};2;{r};{g};{b}m"),
            None => String::new(),
        }
    }
}

fn parse_hex(hex: &[u8]) -> Option<[u8; 3]> {
    let raw = hex.strip_prefix(b"#").unwrap_or(hex);
    if raw.len() != 6 || !raw.iter().all(u8::is_ascii_hexdigit) {
        return None;
    }
    let digit = |byte: u8| {
        if byte.is_ascii_digit() {
            byte - b'0'
        } else {
            byte.to_ascii_lowercase() - b'a' + 10
        }
    };
    Some([
        digit(raw[0]) * 16 + digit(raw[1]),
        digit(raw[2]) * 16 + digit(raw[3]),
        digit(raw[4]) * 16 + digit(raw[5]),
    ])
}

pub fn list_builtins() -> String {
    "catppuccin-mocha  built-in default\nbase16/base24     supported via diffo themes validate <file>\n".into()
}

const MAX_THEME_BYTES: u64 = 4 * 1024 * 1024;

pub fn validate_base_theme_file(path: impl AsRef<Path>) -> Result<String> {
    let file = File::open(path).map_err(|_| Error::ThemeInvalid)?;
    let mut content = Vec::new();
    file.take(MAX_THEME_BYTES + 1)
        .read_to_end(&mut content)
        .map_err(|_| Error::ThemeInvalid)?;
    if content.len() as u64 > MAX_THEME_BYTES {
        return Err(Error::ThemeInvalid);
    }
    validate_base_theme(&content)
}

fn validate_base_theme(content: &[u8]) -> Result<String> {
    let mut base16_count = 0;
    let mut base24_count = 0;
    for index in 0..24 {
        let key = format!("base{index:02x}");
        if key_has_hex_color(content, key.as_bytes()) {
            if index < 16 {
                base16_count += 1;
            } else {
                base24_count += 1;
            }
        }
    }
    if base16_count < 16 {
        return Err(Error::ThemeInvalid);
    }
    Ok(format!(
        "valid Base{} theme ({} color slots)",
        if base24_count >= 8 { "24" } else { "16" },
        base16_count + base24_count
    ))
}

fn key_has_hex_color(content: &[u8], key: &[u8]) -> bool {
    // Match the original format-agnostic validator: a baseNN occurrence must
    // have a standalone six-hex value later on the same line. This is not YAML parsing.
    content
        .windows(key.len())
        .enumerate()
        .any(|(index, window)| {
            if window != key {
                return false;
            }
            let after = &content[index + key.len()..];
            let line_end = after
                .iter()
                .position(|byte| *byte == b'\n')
                .unwrap_or(after.len());
            line_contains_hex_color(&after[..line_end])
        })
}

fn line_contains_hex_color(line: &[u8]) -> bool {
    line.windows(6).enumerate().any(|(index, hex)| {
        let before_ok = index == 0 || !line[index - 1].is_ascii_hexdigit();
        let after_ok = index + 6 == line.len() || !line[index + 6].is_ascii_hexdigit();
        before_ok && after_ok && parse_hex(hex).is_some()
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    fn theme_content(slots: usize, value: &str) -> String {
        (0..slots)
            .map(|index| format!("base{index:02x}: \"{value}\"\n"))
            .collect()
    }

    #[test]
    fn base16_validator_accepts_keys() {
        let mut file = tempfile::NamedTempFile::new().unwrap();
        file.write_all(theme_content(16, "ffffff").as_bytes())
            .unwrap();
        assert_eq!(
            validate_base_theme_file(file.path()).unwrap(),
            "valid Base16 theme (16 color slots)"
        );
    }

    #[test]
    fn validator_recognizes_base24_and_partial_extra_slots() {
        for (slots, name) in [(16, "16"), (19, "16"), (24, "24")] {
            let content = theme_content(slots, "#Aa09Ff");
            assert_eq!(
                validate_base_theme(content.as_bytes()).unwrap(),
                format!("valid Base{name} theme ({slots} color slots)")
            );
        }
    }

    #[test]
    fn validator_rejects_missing_invalid_long_and_cross_line_colors() {
        for content in [
            String::new(),
            theme_content(15, "ffffff"),
            theme_content(16, "fffff"),
            theme_content(16, "fffffff"),
            theme_content(16, "gggggg"),
            theme_content(16, ""),
            theme_content(16, "\nffffff"),
            "base00: ffffff\n".repeat(16),
        ] {
            assert!(
                matches!(
                    validate_base_theme(content.as_bytes()),
                    Err(Error::ThemeInvalid)
                ),
                "{content:?}"
            );
        }
    }

    #[test]
    fn validator_accepts_later_valid_duplicate_and_crlf() {
        let content = format!(
            "{}{}",
            theme_content(16, "invalid"),
            theme_content(16, "123ABC").replace('\n', "\r\n")
        );
        assert_eq!(
            validate_base_theme(content.as_bytes()).unwrap(),
            "valid Base16 theme (16 color slots)"
        );
    }

    #[test]
    fn validator_rejects_unreadable_and_oversized_files() {
        let dir = tempfile::tempdir().unwrap();
        assert!(matches!(
            validate_base_theme_file(dir.path().join("missing.yaml")),
            Err(Error::ThemeInvalid)
        ));
        let mut file = tempfile::NamedTempFile::new_in(dir.path()).unwrap();
        file.write_all(theme_content(16, "ffffff").as_bytes())
            .unwrap();
        file.as_file().set_len(MAX_THEME_BYTES + 1).unwrap();
        assert!(matches!(
            validate_base_theme_file(file.path()),
            Err(Error::ThemeInvalid)
        ));
    }

    #[test]
    fn parse_hex_accepts_optional_hash_and_rejects_malformed_values() {
        assert_eq!(parse_hex(b"#1e1e2e"), Some([30, 30, 46]));
        assert_eq!(parse_hex(b"Aa00FF"), Some([170, 0, 255]));
        for invalid in ["", "#fff", "0000000", "gg0000", "++0000", "架架"] {
            assert_eq!(parse_hex(invalid.as_bytes()), None);
        }
    }

    #[test]
    fn ansi_obeys_no_color_force_and_truecolor_detection_without_global_mutation() {
        assert_eq!(
            Ansi::from_environment(false, true, Some("truecolor")),
            Ansi {
                enabled: false,
                true_color: false
            }
        );
        assert_eq!(
            Ansi::from_environment(true, true, Some("truecolor")),
            Ansi {
                enabled: true,
                true_color: true
            }
        );
        assert_eq!(
            Ansi::from_environment(false, false, None),
            Ansi {
                enabled: true,
                true_color: false
            }
        );
        for value in ["truecolor", "24bit", "supports-truecolor"] {
            assert!(Ansi::from_environment(false, false, Some(value)).true_color);
        }
        assert!(!Ansi::from_environment(false, false, Some("TRUECOLOR")).true_color);
    }

    #[test]
    fn ansi_encodes_rgb_and_handles_disabled_or_invalid_colors() {
        let ansi = Ansi {
            enabled: true,
            true_color: true,
        };
        let color = Color { hex: "#1e1e2e" };
        assert_eq!(ansi.fg(color), "\x1b[38;2;30;30;46m");
        assert_eq!(ansi.bg(color), "\x1b[48;2;30;30;46m");
        assert_eq!(ansi.reset(), "\x1b[0m");
        assert_eq!(ansi.fg(Color { hex: "invalid" }), "");
        for disabled in [
            Ansi {
                enabled: false,
                true_color: true,
            },
            Ansi {
                enabled: true,
                true_color: false,
            },
        ] {
            assert_eq!(disabled.fg(color), "");
            assert_eq!(disabled.bg(color), "");
        }
        assert_eq!(
            Ansi {
                enabled: false,
                true_color: false
            }
            .reset(),
            ""
        );
    }

    #[test]
    fn default_palette_and_builtin_listing_match_original() {
        let palette = catppuccin_mocha();
        assert_eq!(palette.name, "catppuccin-mocha");
        assert_eq!(palette.bg_default.hex, "#1e1e2e");
        assert_eq!(palette.syntax_plain, palette.fg_default);
        assert_eq!(palette.syntax_function, palette.fg_accent);
        assert_eq!(palette.reviewed_badge, palette.diff_add_fg);
        assert_eq!(palette.error_color, palette.diff_del_fg);
        assert_eq!(
            list_builtins(),
            "catppuccin-mocha  built-in default\nbase16/base24     supported via diffo themes validate <file>\n"
        );
    }
}
